### PR TITLE
containerapps: fixing inconsistencies in the swagger

### PR DESCRIPTION
**Pull request by tombuildsstuff**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [x] update existing version to fix swagger quality issue
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/app/resource-manager/Microsoft.App/preview/2022-01-01-preview/ContainerApps.json [hunk 1]
```diff
@@ -427,8 +427,8 @@
         "activeRevisionsMode": {
           "description": "ActiveRevisionsMode controls how active revisions are handled for the Container app:\n<list><item>Multiple: multiple revisions can be active. If no value if provided, this is the default</item><item>Single: Only one revision can be active at a time. Revision weights can not be used in this mode</item></list>",
           "enum": [
-            "multiple",
-            "single"
+            "Multiple",
+            "Single"
```
Comment (tombuildsstuff) [inline specification/app/resource-manager/Microsoft.App/preview/2022-01-01-preview/ContainerApps.json:431]: in this case, the API returns these in TitleCase:

```
{
    "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/stedev-ca-regtest/providers/Microsoft.App/containerApps/stedev-ca-regtest",
    "name": "stedev-ca-regtest",
    "type": "Microsoft.App/containerApps",
    "location": "West Europe",
    "systemData": {
        "createdBy": "XXX",
        "createdByType": "User",
        "createdAt": "2022-04-08T10:38:17.4255847",
        "lastModifiedBy": "XXX",
        "lastModifiedByType": "User",
        "lastModifiedAt": "2022-04-08T10:38:17.4255847"
    },
    "properties": {
        "provisioningState": "Succeeded",
        "managedEnvironmentId": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/stedev-ca-regtest/providers/Microsoft.App/managedenvironments/stedev-ca-regtest",
        "outboundIpAddresses": [
            "20.31.240.94",
            "20.31.240.109",
            "20.31.240.114"
        ],
        "latestRevisionName": "stedev-ca-regtest--bkb6i1h",
        "latestRevisionFqdn": "stedev-ca-regtest--bkb6i1h.ashycoast-1ca0508f.westeurope.azurecontainerapps.io",
        "customDomainVerificationId": "XXXXXXX",
        "configuration": {
            "activeRevisionsMode": "Single",
            "ingress": {
                "fqdn": "stedev-ca-regtest.ashycoast-1ca0508f.westeurope.azurecontainerapps.io",
                "external": true,
                "targetPort": 80,
                "transport": "Auto",
                "traffic": [
                    {
                        "weight": 100,
                        "latestRevision": true
                    }
                ],
                "allowInsecure": false
            },
            "registries": []
        },
        "template": {
            "containers": [
                {
                    "image": "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest",
                    "name": "simple-hello-world-container",
                    "command": [],
                    "resources": {
                        "cpu": 0.25,
                        "memory": "0.5Gi"
                    }
                }
            ],
            "scale": {
                "maxReplicas": 10
            }
        }
    },
    "identity": {
        "type": "None"
    }
}
```

File: specification/app/resource-manager/Microsoft.App/stable/2022-03-01/ContainerApps.json [hunk 2]
```diff
@@ -425,8 +425,8 @@
         "activeRevisionsMode": {
           "description": "ActiveRevisionsMode controls how active revisions are handled for the Container app:\n<list><item>Multiple: multiple revisions can be active.</item><item>Single: Only one revision can be active at a time. Revision weights can not be used in this mode. If no value if provided, this is the default.</item></list>",
           "enum": [
-            "multiple",
-            "single"
+            "Multiple",
+            "Single"
```
Comment (tombuildsstuff) [inline specification/app/resource-manager/Microsoft.App/stable/2022-03-01/ContainerApps.json:429]: in this case, the API returns these in TitleCase:

```
{
    "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/stedev-ca-regtest/providers/Microsoft.App/containerApps/stedev-ca-regtest",
    "name": "stedev-ca-regtest",
    "type": "Microsoft.App/containerApps",
    "location": "West Europe",
    "systemData": {
        "createdBy": "XXX",
        "createdByType": "User",
        "createdAt": "2022-04-08T10:38:17.4255847",
        "lastModifiedBy": "XXX",
        "lastModifiedByType": "User",
        "lastModifiedAt": "2022-04-08T10:38:17.4255847"
    },
    "properties": {
        "provisioningState": "Succeeded",
        "managedEnvironmentId": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/stedev-ca-regtest/providers/Microsoft.App/managedenvironments/stedev-ca-regtest",
        "outboundIpAddresses": [
            "20.31.240.94",
            "20.31.240.109",
            "20.31.240.114"
        ],
        "latestRevisionName": "stedev-ca-regtest--bkb6i1h",
        "latestRevisionFqdn": "stedev-ca-regtest--bkb6i1h.ashycoast-1ca0508f.westeurope.azurecontainerapps.io",
        "customDomainVerificationId": "XXXXXXX",
        "configuration": {
            "activeRevisionsMode": "Single",
            "ingress": {
                "fqdn": "stedev-ca-regtest.ashycoast-1ca0508f.westeurope.azurecontainerapps.io",
                "external": true,
                "targetPort": 80,
                "transport": "Auto",
                "traffic": [
                    {
                        "weight": 100,
                        "latestRevision": true
                    }
                ],
                "allowInsecure": false
            },
            "registries": []
        },
        "template": {
            "containers": [
                {
                    "image": "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest",
                    "name": "simple-hello-world-container",
                    "command": [],
                    "resources": {
                        "cpu": 0.25,
                        "memory": "0.5Gi"
                    }
                }
            ],
            "scale": {
                "maxReplicas": 10
            }
        }
    },
    "identity": {
        "type": "None"
    }
}
```

**Comment by tombuildsstuff** [comment]
@JeffreyRichter a thought which came to mind when putting this together - have you considered adding a linter to ensure these method parameters are consistent? This seems like something which should be consistent across all usages of a given Resource?

**Comment by JeffreyRichter** [comment]
@tombuildsstuff I don't have enough context on this PR. Do you want the linter tool to check casing for consistency or do you want consistent words (multiple/single) to be used across services? I think the former. Azure's guidelines are to use camel casing. I would have thought the linter already checks this. Perhaps @mikekistler can look into this.

**Review by mikekistler — CHANGES_REQUESTED** [review]
I do not think you should be making any changes to the 2022-01-01-preview API. Since there is now a stable 2022-03-01 api version, the preview should be retired 90 days after the stable is available, so there is little value in making changes at this point.

As for the changes to 2022-03-01, these will result in breaking changes to SDKs generated from this API definition, so could break customers.  We need to avoid that unless _absolutely necessary_, and "inconsistency" does not meet that bar.  Can you give an explanation for why this change is _absolutely necessary_?

**Comment by tombuildsstuff** [comment]
@JeffreyRichter 

> @tombuildsstuff I don't have enough context on this PR. Do you want the linter tool to check casing for consistency or do you want consistent words (multiple/single) to be used across services. ? I think the former. Azure's guidelines are to use camel casing. I would have thought the linter already checks this. Perhaps @mikekistler can look into this.

I'm referring to the segments within the URI for user specified segments - if multiple resources are operating on the same URI those segments should be consistently named and cased, for example consider the following endpoints:

> /apps/{name}
> /apps/{appName}/keys/{name}
> /apps/{appName}/keys/{keyName}/rotate
> /apps/{applicationName}/listKeys

Whilst these URIs are all valid in the Swagger, these inconsistencies lead to inconsistencies within the SDKs etc where these are used as method parameters. Instead if these were all consistent:

> /apps/{appName}
> /apps/{appName}/keys/{keyName}
> /apps/{appName}/keys/{keyName}/rotate
> /apps/{appName}/listKeys

When the methods are generated for those, users have a consistent parameter name for a given value - which has UX benefits.

As such I'm proposing a linter to ensure that where multiple Swagger Resources are operating on the same Resource, that the linter validates that the same parameter name (and casing, as you've mentioned these should be `camelCase`) is used across all of them. This would both make the UX consistent as defined above, and in our case (as we're using the Swagger Endpoints as Identifiers) remove the perpetual diff's we're seeing as in this instance.

---

@mikekistler 

Per this comment/API response - the Swagger doesn't match the API definition, so this isn't correctly defined for both API's either way: https://github.com/Azure/azure-rest-api-specs/pull/18660#discussion_r848289233

>  I do not think you should be making any changes to the 2022-01-01-preview API. 

We can revert these if needed, that's fine - at this point in time only the Preview API is used in SDK's (AFAIK) - the Stable API doesn't appear to be rolled out either to Azure Public or the SDKs.

> Since there is now a stable 2022-03-01 api version, the preview should be retired 90 days after the stable is available, so there is little value in making changes at this point.

Is this a new policy? There's many Azure API's which don't comply with this right now (SQL comes to mind, which due to using a composite API version still makes use of Preview functionality from 2015 iirc).

> As for the changes to 2022-03-01, these will result in breaking changes to SDKs generated from this API definition, so could break customers. We need to avoid that unless absolutely necessary, and "inconsistency" does not meet that bar. Can you give an explanation for why this change is absolutely necessary?


There's two sets of changes included for each API version here:

1. Fixing the values returned by the API for the constants
2. Making each of the Resource URIs consistent (as explained in the first half of this comment)

In the case of updating the values used for the constant - these are the actual values returned from the API, see this comment for an example response payload: XXX. If needs be we can revert this for now, but given this is what the API is returning (and that AFAIK the Stable API isn't rolled out to SDKs yet) - I don't believe this is a breaking change at this point in time (for the Stable API, at least)?

For the other - as mentioned above we're making use of the Resource URI as an Identifier, which gets transformed into an object (in our case a Go struct), which we then pass into SDK methods rather than the individual components. This has the UX benefit of highlighting which resource will be modified by which operation, verses outputting the individual fields - and also reduces the amount of code needed (since you can build up the object one time, then create/update/retrieve/delete it as needed).

As it stands right now these URIs being inconsistent means that we're seeing the following perpetual diff:

Screenshot 2022-04-13 at 09 40 54

This means that we're unable to support this resource within Terraform due to the perpetual changes - as such this PR fixes this consistency issue so that we can support this.

As mentioned in the first half of this comment, I think there's UX benefits for end-users in having these consistent (since this'll lead to consistent parameter names in the SDKs etc) - so I'm wondering if there's potentially a missing linter for this too?

Hope that helps clarify that - but please let us know if there's anything else you need to get this merged.

**Comment by mikekistler** [comment]
On the specific topic of the linter rule, you are right that the names should be consistent and the linter can/should check for this.  There is already an issue open to add this linter rule to the standard PR checks:

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1424461

I added a comment there to hopefully raise the priority on this.

And it is already implemented in the [Azure API Style Guide](https://github.com/Azure/azure-api-style-guide) linter rules -- here's a screen cap of the linter running in VSCode flagging this particular issue:

image

So in short, we agree on using consistent names!

**Comment by mikekistler** [comment]
On the other points in this PR:

- Changing the enum values to match the service.  It is regrettable that this was not caught earlier but changing the API def to match the service is generally an approved breaking change.

- We can revert these [the changes to the preview version] if needed.  Yes please.

- The 2022-03-01 stable is not rolled out. This isn't a great situation. Our PR automation assumes that a "stable" API that is merged to main in this repo is considered "released". But if this truly is the case, then this might make the path param name changes acceptable.

- The preview should be retired 90 days after the stable is available.  This is not a new policy but I'll grant you that it needs to be more clearly/broadly communicated and we know many services are not in compliance.  We can't fix everything all at once ;-).

tl;dr Keep the constant name changes and path param changes in the stable version only, come to the Breaking Changes office hours and plead your case that the API is not yet rolled out, and there's a fair chance this will be approved.

**Comment by ruslany** [comment]
@mikekistler , @tombuildsstuff - both 2022-03-01 and 2022-01-01-preview API versions have been rolled out in production, but the SDK clients haven't been released for any of those versions. Because of that I think it should be ok to merge this PR as it now matches the casing to what RP actually returns.

Thanks @tombuildsstuff  for creating this PR - it is unfortunate we haven't caught the casing of the enum values earlier.

**Comment by mikekistler** [comment]
Since this PR is flagged as "BreakingChangeReviewRequired", please open a breaking change intake and come to the next Breaking Change office hours (Mondays 10-11 PT). I don't expect any issues with getting this approved but it should go through the process.

**Comment by ruslany** [comment]
@mikekistler - our team also create this PR to fix a few other case inconsistencies for the enum values. https://github.com/Azure/azure-rest-api-specs/pull/18839

**Comment by ruslany** [comment]
@tombuildsstuff - can you please update the PR to be in sync with the latest state of the main branch?

**Comment by tombuildsstuff** [comment]
Closing [as per this comment](https://github.com/Azure/azure-rest-api-specs/pull/19005#issuecomment-1124035417) since this has been merged via main - thanks both.
